Source/WebCore/page/PointerLockController.h

```diff
@@ -65,13 +73,17 @@ class PointerLockController {
     void clearElement();
     void enqueueEvent(const AtomString& type, Element*);
     void enqueueEvent(const AtomString& type, Document*);
+    void resolvePromises();
+    void rejectPromises(ExceptionCode, const String&);
     void elementWasRemovedInternal();
 
     Page& m_page;
     bool m_lockPending { false };
     bool m_unlockPending { false };
     bool m_forceCursorVisibleUponUnlock { false };
+    std::optional<PointerLockOptions> m_options;
     RefPtr<Element> m_element;
+    Vector<Ref<DeferredPromise>> m_promises;
```

Comment:
Think about whether you're introducing any retain cycles here (I don't know if you are).

Reply:
Good question. It motivated me to write this quick test page:

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <title></title>
    <script>
        async function lock()
        {
            const target = document.getElementById('target');
            console.log("locking", target);
            let promise = target.requestPointerLock();
            target.remove();
            try {
                await promise;
                console.log("Promise resolved");
            } catch (err) {
                console.error("Promise rejected", err);
            }
        }
    </script>
</head>
<body>
    <div id="target" style="width: 200px; height: 200px; background-color: purple;"></div>
    <button onclick="lock()">Lock it</button>
</body>
</html>
```

And when I run it I crash here:

```
void PointerLockController::clearElement()
{
    m_lockPending = false;
    m_element = nullptr;
    m_options = std::nullopt;
    ASSERT(m_promises.isEmpty()); // <-- assertion hit
}
```

I will add a test for this and also fix the bug in my next push. Similarly, an assertion is also possible if `location.reload()` is used instead of `target.remove()` and completionHandler from `- (void)_webViewDidRequestPointerLock:(WKWebView *)webView completionHandler:(void (^)(BOOL))completionHandler` isn't called immediately (MiniBrowser does, but I don't know what Safari does and the method signature certainly hints that delegates can take their time in resolving completionHandler).

Reply:
Test added and bug fixed in my latest push.